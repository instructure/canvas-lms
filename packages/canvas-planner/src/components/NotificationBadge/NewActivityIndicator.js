/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {Component} from 'react'
import {arrayOf, string, number, func} from 'prop-types'
import formatMessage from '../../format-message'

import {animatable} from '../../dynamic-ui'

import Indicator from './Indicator'

export class NewActivityIndicator extends Component {
  static propTypes = {
    title: string.isRequired,
    itemIds: arrayOf(string).isRequired,
    registerAnimatable: func,
    deregisterAnimatable: func,
    animatableIndex: number,
    getFocusable: func
  }

  static defaultProps = {
    registerAnimatable: () => {},
    deregisterAnimatable: () => {}
  }

  UNSAFE_componentWillMount() {
    this.props.registerAnimatable(
      'new-activity-indicator',
      this,
      this.props.animatableIndex,
      this.props.itemIds
    )
  }

  UNSAFE_componentWillReceiveProps(newProps) {
    this.props.deregisterAnimatable('new-activity-indicator', this, this.props.itemIds)
    this.props.registerAnimatable(
      'new-activity-indicator',
      this,
      newProps.animatableIndex,
      newProps.itemIds
    )
  }

  componentWillUnmount() {
    this.props.deregisterAnimatable('new-activity-indicator', this, this.props.itemIds)
  }

  getFocusable() {
    return this.props.getFocusable ? this.props.getFocusable() : undefined
  }

  getScrollable() {
    return this.indicatorElt
  }

  render() {
    const badgeMessage = formatMessage('New activity for {title}', {title: this.props.title})
    return (
      <Indicator
        indicatorRef={ref => (this.indicatorElt = ref)}
        title={badgeMessage}
        variant="primary"
      />
    )
  }
}

export default animatable(NewActivityIndicator)
