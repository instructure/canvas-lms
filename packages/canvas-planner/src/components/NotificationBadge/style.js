/*
 * Copyright (C) 2023 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {getThemeVars} from '../../getThemeVars'

export default function buildStyle() {
  /*
   * If the theme variables to be used when generating the styles below
   * are dependent on the actual theme in use, you can also pull out the
   * `key` property from the return from `getThemeVars()` and do a bit of
   * if or switch statement logic to get the result you want.
   */
  const {variables} = getThemeVars()

  const classNames = {
    activityIndicator: 'NotificationBadge-styles__activityIndicator',
    hasBadge: 'NotificationBadge-styles__hasBadge',
    small: 'NotificationBadge-styles__small',
  }

  const theme = {
    activityIndicatorPadding: variables.spacing.small,
    activityIndicatorWidth: variables.spacing.small,
    activityIndicatorBorderSize: '1rem',
    activityIndicatorBackground: variables.colors.white,
  }

  const css = `
  .${classNames.activityIndicator} {
    width: ${theme.activityIndicatorWidth};
    padding: ${theme.activityIndicatorPadding};
  }
  .${classNames.activityIndicator}.${classNames.hasBadge} {
    background: transparent;
    width: auto;
    height: auto;
    align-items: center;
    justify-content: center;
    position: static;
    display: flex;
    top: auto;
    right: auto;
    z-index: 1;
    border-radius: 0;
  }
  
  .${classNames.small}.${classNames.activityIndicator} {
    padding: 0;
  }
  `

  return {css, classNames, theme}
}
