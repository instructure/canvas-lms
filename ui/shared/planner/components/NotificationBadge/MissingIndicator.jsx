/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {string} from 'prop-types'
import {useScope as createI18nScope} from '@canvas/i18n'
import Indicator from './Indicator'

const I18n = createI18nScope('planner')

export default function MissingIndicator(props) {
  const badgeMessage = I18n.t('Missing items for %{title}', {title: props.title})
  return (
    <Indicator
      title={badgeMessage}
      variant="invisible"
      testId={props.testId || 'missing-indicator'}
    />
  )
}

MissingIndicator.propTypes = {
  title: string.isRequired,
  testId: string,
}
