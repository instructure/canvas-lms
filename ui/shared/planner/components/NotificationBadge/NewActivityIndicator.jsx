/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React, {Component} from 'react'
import {arrayOf, string, number, func} from 'prop-types'
import {useScope as createI18nScope} from '@canvas/i18n'
import {animatable} from '../../dynamic-ui'
import Indicator from './Indicator'

const I18n = createI18nScope('planner')

export class NewActivityIndicator extends Component {
  static propTypes = {
    title: string.isRequired,
    itemIds: arrayOf(string).isRequired,
    registerAnimatable: func,
    deregisterAnimatable: func,
    animatableIndex: number,
    getFocusable: func,
    testId: string,
  }

  static defaultProps = {
    registerAnimatable: () => {},
    deregisterAnimatable: () => {},
  }

  UNSAFE_componentWillMount() {
    this.props.registerAnimatable(
      'new-activity-indicator',
      this,
      this.props.animatableIndex,
      this.props.itemIds,
    )
  }

  UNSAFE_componentWillReceiveProps(newProps) {
    this.props.deregisterAnimatable('new-activity-indicator', this, this.props.itemIds)
    this.props.registerAnimatable(
      'new-activity-indicator',
      this,
      newProps.animatableIndex,
      newProps.itemIds,
    )
  }

  componentWillUnmount() {
    this.props.deregisterAnimatable('new-activity-indicator', this, this.props.itemIds)
  }

  getFocusable() {
    return this.props.getFocusable ? this.props.getFocusable() : undefined
  }

  getScrollable() {
    return this.indicatorElt
  }

  render() {
    const badgeMessage = I18n.t('New activity for %{title}', {title: this.props.title})
    return (
      <Indicator
        indicatorRef={ref => (this.indicatorElt = ref)}
        title={badgeMessage}
        variant="primary"
        testId={this.props.testId || 'new-activity-indicator'}
      />
    )
  }
}

NewActivityIndicator.displayName = 'NewActivityIndicator'

export default animatable(NewActivityIndicator)
