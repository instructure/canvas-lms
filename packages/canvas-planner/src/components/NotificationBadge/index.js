/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {themeable} from '@instructure/ui-themeable'
import MissingIndicator from './MissingIndicator'
import NewActivityIndicator from './NewActivityIndicator'
import {sizeShape} from '../plannerPropTypes'
import styles from './styles.css'
import theme from './theme'

class NotificationBadge extends React.Component {
  static propTypes = {
    responsiveSize: sizeShape,
    children: PropTypes.element
  }

  static defaultProps = {
    responsiveSize: 'large'
  }

  render() {
    const indicator = this.props.children ? React.Children.only(this.props.children) : null

    const activityIndicatorClasses = classnames(
      styles.activityIndicator,
      indicator != null && styles.hasBadge,
      styles[this.props.responsiveSize]
    )

    return <div className={activityIndicatorClasses}>{indicator}</div>
  }
}
const ThemeableNotificationBadge = themeable(theme, styles)(NotificationBadge)

export {MissingIndicator, NewActivityIndicator, NotificationBadge}
export default ThemeableNotificationBadge
