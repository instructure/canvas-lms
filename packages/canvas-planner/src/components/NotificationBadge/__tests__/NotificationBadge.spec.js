/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {shallow} from 'enzyme'
import NotificationBadge, {NewActivityIndicator} from '../index'

// it would be better if the snapshots contained the proper class names, but
// jest doesn't deal with how themeable turns styles.css into code.
it('renders an indicator', () => {
  const wrapper = shallow(
    <NotificationBadge>
      <NewActivityIndicator title="blah" />
    </NotificationBadge>
  )
  expect(wrapper).toMatchSnapshot()
})

it('renders an empty div', () => {
  const wrapper = shallow(<NotificationBadge>{null}</NotificationBadge>)
  expect(wrapper).toMatchSnapshot()
})
