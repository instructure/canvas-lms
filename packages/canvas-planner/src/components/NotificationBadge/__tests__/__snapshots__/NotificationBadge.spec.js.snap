// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty div 1`] = `
<div
  className="NotificationBadge-styles__activityIndicator"
/>
`;

exports[`renders an indicator 1`] = `
<div
  className="NotificationBadge-styles__activityIndicator NotificationBadge-styles__hasBadge"
>
  <Animatable(NewActivityIndicator)
    title="blah"
  />
</div>
`;
